File: lc_soln/0098_valid_binary_search_tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root, float("-inf"), float("inf"))

    def dfs(self, root: Optional[TreeNode], min_val: int, max_val: int):
        if root is None:
            return True
        if not (min_val < root.val < max_val):
            return False

        return self.dfs(root.left, min_val, root.val) and self.dfs(
            root.right, root.val, max_val
        )


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0098_valid_binary_search_tree/solution.rs
struct Solution;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::dfs(root, i32::MIN, i32::MAX)
    }

    fn dfs(root: Option<Rc<RefCell<TreeNode>>>, min_val: i32, max_val: i32) -> bool {
        match root {
            None => true,
            Some(r) => {
                let node = r.borrow();
                if !(min_val < node.val && node.val < max_val) {
                    return false;
                }
                Self::dfs(node.left.clone(), min_val, node.val)
                    && Self::dfs(node.right.clone(), node.val, max_val)
            }
        }
    }
}

fn main() {}
